httpserver/service_node.cpp

```diff
@@ -26,6 +26,50 @@ using service_node::storage::Item;
 namespace loki {
 using http_server::connection_t;
 
+constexpr std::array<std::chrono::seconds, 5> RETRY_INTERVALS = {
+    std::chrono::seconds(5), std::chrono::seconds(10), std::chrono::seconds(20),
+    std::chrono::seconds(40), std::chrono::seconds(80)};
+
+FailedWork::FailedWork(boost::asio::io_context& ioc, const sn_record_t& sn,
+                       std::shared_ptr<request_t> req)
+    : ioc_(ioc), retry_timer_(ioc), sn_(sn), request_(std::move(req)) {}
+
+void FailedWork::retry(std::shared_ptr<FailedWork>&& self) {
+
+    attempt_count_ += 1;
+    if (attempt_count_ > RETRY_INTERVALS.size()) {
+        BOOST_LOG_TRIVIAL(debug)
+            << "Gave up after " << attempt_count_ << " attempts";
+        return;
+    }
+
+    retry_timer_.expires_after(RETRY_INTERVALS[attempt_count_ - 1]);
+    BOOST_LOG_TRIVIAL(debug)
+        << "Will retry in " << RETRY_INTERVALS[attempt_count_ - 1].count()
+        << " secs";
+
+    retry_timer_.async_wait(
+        [self = std::move(self)](const boost::system::error_code& ec) mutable {
+            const auto& sn = self->sn_;
+
+            make_http_request(
+                self->ioc_, sn.address, sn.port, *self->request_,
```

Comment:
Maybe best take references to `ioc_` and `request_` before calling `make_http_request` to prevent trying to dereference `self` after it has been moved...Tricky!

Reply:
updated